Add documentation for event filters

Deamon5550 opened the pull request:
This adds documentation for [SpongeAPI#927 Event Filters](https://github.com/SpongePowered/SpongeAPI/pull/927)

Please feel free to edit this branch for content/formatting/whatever without asking me for permission.

Deamon5550 committed:
Start documentation for event filters

Spongy commented:
A preview for this pull request is available at https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/f4efcebe/index.html.

Here are some links to the pages that were modified:

- added https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/f4efcebe/plugin/events/event-filters.html

_Since the preview frequently changes, please link to [this comment](https://github.com/SpongePowered/SpongeDocs/pull/379#issuecomment-160007573), not to the direct url to the preview._

Deamon5550 committed:
Add event-filters to index

Saladoc committed:
Split Events page in several sections

- Stub for Causes page
- consistent indentation (tabs -> spaces in filters.rst, 4 spaces per
  level in custom.rst)

Deamon5550 commented on source/plugin/event/event-filters.rst:
```diff
@@ -0,0 +1,145 @@
+======================================
+Advansed Event Handling: Event Filters
```
Should probably change this to just 'Event Filters'

Saladoc committed:
Fix the reference on main class page

saladoc commented:
Open questions of lesser priority: 
 - Can event filters be applied to event listeners registered as per https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e1a2a6ff/plugin/event/listeners.html#dynamically-registering-event-listeners ? If so, how?
 - `@Listener(ignoreCancelled=false)` vs `@Listener @IsCancelled(CancellationState.IGNORE)` - which one should be preferred?

Deamon5550 commented:
> - Can event filters be applied to event listeners registered as per https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e1a2a6ff/plugin/event/listeners.html#dynamically-registering-event-listeners ? If so, how?

No they cannot, I toyed briefly with how to do it but there isn't really a way for a) them to cleanly add the filters to their handler since they are implementing a method that they can't add parameters to, and b) us to cleanly generate/transform their handlers to do the filtering. So bottom line, if you're implementing EventListener yourself you're on your own for filtering.

 > - `@Listener(ignoreCancelled=false)` vs `@Listener @IsCancelled(CancellationState.IGNORE)` - which one should be preferred?

`@Listener(ignoreCancelled=false)` is going to be removed soon and `@IsCancelled` will be the only place for controlling behavior around cancellation. I'll update this once I make that change.

Tzky commented on source/plugin/event/index.rst:
```diff
@@ -0,0 +1,39 @@
+- Listen for events
+- Fire events
+
+Overview
```
I'd say we should get rid of the ``Overview`` heading and the list above. Just some introductory sentences should suffice.

Tzky commented:
Is the concept of our events explained anywhere (``Targets``+``Causes``)? Same for the naming scheme...

saladoc commented:
The events page we have mostly deals with listeners. Causes I introduced as a stub page, thats #267.

Deamon5550 commented:
I'll write a page for causes sometime today

Deamon5550 committed:
First pass at some documentation of Cause

saladoc commented:
Looks good so far. Maybe mention that `NamedCause` also holds constants holding common cause names so you don't have to toss around magic values.

Deamon5550 commented:
@saladoc I added a note to the bottom of the named cause section https://github.com/SpongePowered/SpongeDocs/pull/379/files#diff-dba86822fc60edfa471b8efba4f5b7e2R85

Inscrutable commented on source/plugin/event/causes.rst:
```diff
@@ -0,0 +1,121 @@
+Events are great for the use case of wanting to execute something when certain actions occur, but
+they have the drawback of providing next to no context as to what has **caused** that event to
+occur. A ``Cause`` allows providing and receiving additional contextual information about the event
+which can be do to modify the behavior of your event handler depending on this context.
```
Something is wrong with this sentence, "which can be do to modify". We can clean up some wording later, but I'm really not sure what you meant to say here.

Deamon5550 replied:
I think I meant to say `...which can be used to modify...`

Inscrutable commented on source/plugin/event/causes.rst:
```diff
@@ -0,0 +1,121 @@
+Retrieving objects from a Cause
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Structureally a ``Cause`` object contains a sequential list of objects. There are several methods of
```
"Structurally," (typo and add a comma). In general add a lot more commas.

Inscrutable commented on source/plugin/event/causes.rst:
```diff
@@ -0,0 +1,121 @@
+
+Creating a cause to use when firing an event is extremely easy. The hardest part is deciding
+what information to include in the cause. If you're firing an event from your plugin which is
+usually triggered through other means perhaps you want to include your plguin container so
```
"plugin"

Inscrutable commented on source/plugin/event/causes.rst:
```diff
@@ -0,0 +1,121 @@
+what information to include in the cause. If you're firing an event from your plugin which is
+usually triggered through other means perhaps you want to include your plguin container so
+other plugins know that the event comes from your plugin. Or if you are firing the event on
+behalf of a player due to some action its usually a good idea to include that player in
```
the catastrophe of the missing apostrophe. "it's" (it is).

Deamon5550 committed:
Clean up a few things on the cause page

Deamon5550 committed:
Update event filter page for final event filters

Inscrutable committed:
Fix Dot Points

Deamon5550 committed:
Update listeners page after event filter changes

boformer commented on source/plugin/event/filters.rst:
```diff
@@ -0,0 +1,155 @@
+    }
+
+.. note::
+    Both ``@Has`` and ``@Supports`` have an optional parameter ``inverse()`` which can be set to cause validation
```
maybe remove the brackets?

boformer commented:
Looks good.

Aaron1011 committed:
Document @Before, @After, and @Named

Aaron1011 committed:
Fix ending bold symbols
